samples/java-eventsourced-shopping-cart/src/main/java/com/example/shoppingcart/domain/ShoppingCartImpl.java

```diff
@@ -16,94 +16,111 @@
 
 package com.example.shoppingcart.domain;
 
+import com.akkaserverless.javasdk.Effect;
 import com.akkaserverless.javasdk.EntityId;
 import com.akkaserverless.javasdk.eventsourcedentity.CommandContext;
 import com.akkaserverless.javasdk.eventsourcedentity.EventSourcedEntity;
+import com.akkaserverless.javasdk.eventsourcedentity.EventSourcedEntityEffect;
 import com.example.shoppingcart.ShoppingCartApi;
 import com.google.protobuf.Empty;
 
+import java.util.Comparator;
 import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
 @EventSourcedEntity(entityType = "eventsourced-shopping-cart")
-public class ShoppingCartImpl extends ShoppingCartInterface {
+public class ShoppingCartImpl extends ShoppingCartInterface2 {
   @SuppressWarnings("unused")
   private final String entityId;
 
-  private final Map<String, ShoppingCartApi.LineItem> cart = new LinkedHashMap<>();
-
   public ShoppingCartImpl(@EntityId String entityId) {
     this.entityId = entityId;
   }
 
   @Override
-  public ShoppingCartDomain.Cart snapshot() {
-    return ShoppingCartDomain.Cart.newBuilder()
-        .addAllItems(cart.values().stream().map(this::convert).collect(Collectors.toList()))
-        .build();
+  protected ShoppingCartDomain.Cart emptyState() {
+    return ShoppingCartDomain.Cart.getDefaultInstance();
```

Comment:
I like how explicit this is.

Reply:
If we have `emptyState` we have to include make that accessible from the SDK also, not only in generated code like this. We could have an interface or abstract class that the impl has to implement.

Reply:
@octonato not that I would like to bring back memories to our old discussion about emptyState 😄 , but what do you think we should do here? `emptyState` or `Optional` parameters everywhere? Same in ValueEntity or is the difference motivated?

Reply:
Thanks for bring back those memories! :-)

My whole idea in the past is that I wanted to avoid exposing users to `Option` because I believe there is a clear point in time where the journal is empty and therefore we start with `None`. So, first event moves from `None` to `Some`. All subsequent events can only work on a `Some`. 

From that observation, we could say there are two possible event handlers (and command handlers): `Event => State` and `(State, Event) => State`. And we can then hide the `Option` from the user. 

Anyway, that just to remind my opinion about it since you triggered my memories. :-)

Back to the present...

I think we don't want to introduce two kinds of command and event handlers because that would require the user to annotate the methods in the proto file. Event handlers would even be more challenging because there are not even defined in the proto file. 

So, that to say, my opinion is that we should not bother users with an `Optional` type if most of the time it will be a `Some`. Only the first command handler and event handler will see it as whatever the user defines as empty state. 

And most probably, java developers will use a `null`. 

For the `ValueEntity`, I think we can apply the same reasoning. The `emtpyState` will only be used when there is no row (including after deleting a row and starting over again).

Reply:
I agree. Then I introduce `emptyState` in ValueEntity also to make it consistent.

`null` or the proto empty instance can be used as starting point.

Reply:
Added emptyState for ValueEntity also: bd4aff8